### ISSUE TITLE
$and with $all and $nin not working

### ISSUE THREAD

**Issue by markbigler** [bwaldvogel/mongo-java-server#36]
For the known [reasons](https://github.com/fakemongo/fongo/issues/316#issuecomment-436440615) I have tried to switch from `fongo` to `mongo-java-server`.

Except for the following case, all tests ran correctly. Here the demo test class to reproduce the problem (find all entities that have a tag "A", but not a tag "B" or "C" -> expect no entities at all):

``` java
@RunWith(SpringRunner.class)
@SpringBootTest(classes={Issue.TestConfiguration.class})
public class Issue {

	@Autowired private TestRepository repository;

	@Test
	public void testTags() throws Exception {
		repository.deleteAll();
		repository.insert(new TestEntity("ID_1", "A", "B"));
		repository.insert(new TestEntity("ID_2", "A", "C"));

		List<TestEntity> entitiesByTags = repository.findByTags(Arrays.asList("A"), Arrays.asList("B", "C"));
		assertEquals(0, entitiesByTags.size());
	}

	@Configuration
	@EnableMongoRepositories(basePackageClasses={TestRepository.class})
	protected static class TestConfiguration {
		@Bean
		public MongoTemplate mongoTemplate(MongoClient mongoClient) {
			return new MongoTemplate(mongoDbFactory(mongoClient));
		}

		@Bean
		public MongoDbFactory mongoDbFactory(MongoClient mongoClient) {
			return new SimpleMongoDbFactory(mongoClient, "test");
		}

		@Bean(destroyMethod="shutdown")
		public MongoServer mongoServer() {
			MongoServer mongoServer = new MongoServer(new MemoryBackend());
			mongoServer.bind();
			return mongoServer;
		}

		@Bean(destroyMethod="close")
		public MongoClient mongoClient(MongoServer mongoServer) {
			return new MongoClient(new ServerAddress(mongoServer.getLocalAddress()));
		}
	}
}
```

``` java
@Document(collection="test")
public class TestEntity {
	@Id private String id;
	@Indexed private Set<String> tags = new HashSet<>();
	public TestEntity() {
	}
	public TestEntity(String id, String... tags) {
		this.id = id;
		this.tags = new HashSet<>(Arrays.asList(tags));
	}
	public String getId() {
		return id;
	}
	public Set<String> getTags() {
		return tags;
	}
}
```

``` java
public interface TestRepository extends MongoRepository<TestEntity, String> {
	/**
	 * @param all https://docs.mongodb.com/manual/reference/operator/query/all/
	 * @param nin https://docs.mongodb.com/manual/reference/operator/query/nin/
	 * @return a list of {@link TestEntity} that contains {@code $all} the specified tags but not the tags specified in {@code $nin}.
	 */
	@Query("{$and:[{'tags':{$all:?0}},{'tags':{$nin:?1}}]}")
	public List<TestEntity> findByTags(Collection<String> all, Collection<String> nin);
}
```

**Comment by bwaldvogel** [comment]
Thanks for reporting this bug.

Can you retest with 3d7062d?

**Comment by markbigler** [comment]
Thanks for the fix. I retested it and it works.

However, I found another bug.

I extended the `TestEntity` with a simple string property:
```java
@Document(collection="test")
public class TestEntity {
	@Id private String id;
	private String text;
	@Indexed private Set<String> tags = new HashSet<>();
	public TestEntity() {
	}
	public TestEntity(String id, String text, String... tags) {
		this.id = id;
		this.text = text;
		this.tags = new HashSet<>(Arrays.asList(tags));
	}
	public String getId() {
		return id;
	}
	public String getText() {
		return text;
	}
	public Set<String> getTags() {
		return tags;
	}
}
```

The extended repository:
```java
public interface TestRepository extends MongoRepository<TestEntity, String> {
	@ExistsQuery("{$and:[{'text':?0},{'tags':{$all:?1}}]}")
	public boolean existsByTextAndTags(String text, Collection<String> all);
}
```

And the new test case:
```java
@Test
public void testExistsByTextAndTags() throws Exception {
	repository.deleteAll();
	repository.insert(new TestEntity("ID_1", "TextA", "TagA", "TagB"));
	repository.insert(new TestEntity("ID_2", "TextA", "TagA", "TagC"));
		
	assertTrue(repository.existsByTextAndTags("TextA", Arrays.asList("TagA")));
	assertTrue(repository.existsByTextAndTags("TextA", Arrays.asList("TagA", "TagB")));

	assertFalse(repository.existsByTextAndTags("TextB", Arrays.asList("TagA")));
	assertFalse(repository.existsByTextAndTags("TextA", Arrays.asList())); // fails
}
```

**Comment by markbigler** [comment]
Test case to add to `DefaultQueryMatcherTest`

```java
@Test
public void testMatches() throws Exception {
        Document query = json("$and: [{'text': 'TextA'}, {'tags': {$all: []}}]");

        assertThat(matcher.matches(json("text: 'TextA', tags: []"), query)).isTrue();
        assertThat(matcher.matches(json("text: 'TextB', tags: []"), query)).isFalse();
        assertThat(matcher.matches(json("text: 'TextA', tags: ['A']"), query)).isFalse();
}
```

**Comment by bwaldvogel** [comment]
Thanks.

However, actually I observe a slightly different behavior on a real MongoDB which *never* seems to match for empty lists in `$all` expressions.

The behavior is adjusted accordingly in 70902db. Can you retest?

**Comment by markbigler** [comment]
> However, I observe a slightly different behavior on a real MongoDB.
> It _never_ seems to match for empty lists in `$all` expressions.

You are absolutely correct, `$all:[]` never matches an empty list. Sorry about that.

> Can you retest?

All tests are green now. Thanks alot! 🎁

**Comment by bwaldvogel** [comment]
Fixes are released in version `1.10.0`
